File: packages/FFIJNI/dcf_primitives/ios/Classes/Components/List/DCFScrollable.swift
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import UIKit
import dcflight

/// Custom UIScrollView that implements VirtualizedList content size management
/// Key insight: Yoga handles layout, but contentSize must be explicitly managed
class DCFScrollableView: UIScrollView {
    
    var isHorizontal: Bool = false
    var virtualizedContentOffsetStart: CGFloat = 0
    var virtualizedContentPaddingTop: CGFloat = 0
    var nodeId: String?
    
    private var explicitContentSize: CGSize?
    private var lastFrameSize: CGSize = .zero
    private var isUpdatingContentSize: Bool = false // Prevent redundant calculations
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupDCFScrollableView()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupDCFScrollableView()
    }
    
    private func setupDCFScrollableView() {
        self.clipsToBounds = true
    }
    
    /// Update content size based on Yoga layout results - React Native VirtualizedList approach
    func updateContentSizeFromYogaLayout() {
        guard !isUpdatingContentSize else { return }
        isUpdatingContentSize = true
        defer { isUpdatingContentSize = false }
        
        if let explicitSize = explicitContentSize {
            self.contentSize = explicitSize
            return
        }
        
        var maxWidth: CGFloat = 0
        var maxHeight: CGFloat = 0
        
        for subview in self.subviews {
            let className = NSStringFromClass(type(of: subview))
            guard !className.contains("UIScrollView") && !className.contains("_UIScrollViewScrollIndicator") else { continue }
            
            let right = subview.frame.origin.x + subview.frame.size.width
            let bottom = subview.frame.origin.y + subview.frame.size.height
            
            maxWidth = max(maxWidth, right)
            maxHeight = max(maxHeight, bottom)
        }
        
        if virtualizedContentOffsetStart > 0 || virtualizedContentPaddingTop > 0 {
            let extraPadding = max(virtualizedContentOffsetStart, virtualizedContentPaddingTop)
            
            if isHorizontal {
                maxWidth += extraPadding
                
                for subview in self.subviews {
                    let className = NSStringFromClass(type(of: subview))
                    guard !className.contains("UIScrollView") else { continue }
                    
                    var frame = subview.frame
                    frame.origin.x += extraPadding
                    subview.frame = frame
                }
            } else {
                maxHeight += extraPadding
                
                for subview in self.subviews {
                    let className = NSStringFromClass(type(of: subview))
                    guard !className.contains("UIScrollView") else { continue }
                    
                    var frame = subview.frame
                    frame.origin.y += extraPadding
                    subview.frame = frame
                }
            }
        }
        
        let availableWidth = self.frame.width
        let availableHeight = self.frame.height
        
        let finalContentSize: CGSize
        if isHorizontal {
            finalContentSize = CGSize(width: maxWidth, height: availableHeight)
        } else {
            finalContentSize = CGSize(width: availableWidth, height: maxHeight)
        }
        
        if self.contentSize != finalContentSize {
            self.contentSize = finalContentSize
            
            notifyContentSizeUpdate(finalContentSize)
        }
    }
    
    /// Set explicit content size from Dart side - VirtualizedList approach
    func setExplicitContentSize(_ size: CGSize) {
        explicitContentSize = size
        
        if self.contentSize != size {
            self.contentSize = size
            
            notifyContentSizeUpdate(size)
        }
    }
    
    /// Notify Dart side of content size updates through simple propagateEvent
    private func notifyContentSizeUpdate(_ size: CGSize) {
        propagateEvent(on: self, eventName: "onContentSizeChange", data: [
            "contentSize": [
                "width": size.width,
                "height": size.height
            ]
        ])
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        let currentFrameSize = self.frame.size
        
        if lastFrameSize != currentFrameSize && !isUpdatingContentSize {
            lastFrameSize = currentFrameSize
            
            if currentFrameSize.width > 0 && currentFrameSize.height > 0 && self.subviews.count > 0 {
                // 🚀 CRITICAL: Delay to ensure Yoga layout is complete after orientation change
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
                    guard !self.isUpdatingContentSize else { return }
                    self.updateContentSizeFromYogaLayout()
                }
            }
        }
    }
    
    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        super.traitCollectionDidChange(previousTraitCollection)
        
        if let previous = previousTraitCollection,
           previous.verticalSizeClass != self.traitCollection.verticalSizeClass ||
           previous.horizontalSizeClass != self.traitCollection.horizontalSizeClass {
            explicitContentSize = nil
        }
    }
}

File: packages/dcf_primitives/android/bin/src/main/kotlin/com/dotcorr/dcf_primitives/components/DCFScrollable.kt
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.dotcorr.dcf_primitives.components

import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.core.widget.NestedScrollView
import com.dotcorr.dcflight.components.DCFContentContainerProvider
import com.dotcorr.dcflight.components.DCFTags
import com.dotcorr.dcflight.components.propagateEvent

/**
 * Custom ScrollView that implements VirtualizedList content size management
 * Key insight: Yoga handles layout, but contentSize must be explicitly managed
 * 
 * Matches iOS DCFScrollableView behavior
 */
class DCFScrollableView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : NestedScrollView(context, attrs, defStyleAttr), DCFContentContainerProvider {
    
    companion object {
        private const val TAG = "DCFScrollableView"
    }
    
    var isHorizontal: Boolean = false
    var virtualizedContentOffsetStart: Float = 0f
    var virtualizedContentPaddingTop: Float = 0f
    var nodeId: String? = null
    
    private var explicitContentSize: Pair<Int, Int>? = null
    private var isUpdatingContentSize: Boolean = false // Prevent redundant calculations
    
    // Content container for children (NestedScrollView can only have one direct child)
    private var contentContainer: FrameLayout? = null
    
    init {
        setupDCFScrollableView()
    }
    
    private fun setupDCFScrollableView() {
        clipToPadding = true
        clipChildren = true
        
        // Prevent ScrollView from auto-scrolling to focused child on initial layout
        descendantFocusability = ViewGroup.FOCUS_BEFORE_DESCENDANTS
        isFocusable = true
        isFocusableInTouchMode = true
        
        // Create content container (will be added by component)
        contentContainer = FrameLayout(context)
    }
    
    /**
     * Set the content container (called by DCFScrollViewComponent)
     */
    fun setContentContainer(container: FrameLayout) {
        contentContainer = container
    }
    
    /**
     * DCFContentContainerProvider implementation
     * Returns the FrameLayout content container where children should be attached
     */
    override fun getContentContainer(): ViewGroup? {
        return contentContainer
    }
    
    /**
     * Update content size based on Yoga layout results
     * 
     * Simplified approach:
     * 1. The content container (FrameLayout) holds all children.
     * 2. Children are positioned by Yoga.
     * 3. We just need to ensure the container is large enough to hold them.
     * 4. NestedScrollView automatically handles scrolling if the child is larger than itself.
     */
    fun updateContentSizeFromYogaLayout() {
        if (isUpdatingContentSize) {
            return
        }
        isUpdatingContentSize = true
        
        post {
            try {
                if (explicitContentSize != null) {
                    val (width, height) = explicitContentSize!!
                    setContentSize(width, height)
                    return@post
                }
                
                val contentView = getChildAt(0) as? FrameLayout ?: return@post
                
                // Calculate max bounds from children's actual positions
                var maxWidth = 0
                var maxHeight = 0
                
                for (i in 0 until contentView.childCount) {
                    val child = contentView.getChildAt(i)
                    if (child.visibility != View.GONE) {
                        val right = child.left + child.width
                        val bottom = child.top + child.height
                        maxWidth = maxOf(maxWidth, right)
                        maxHeight = maxOf(maxHeight, bottom)
                    }
                }
                
                // Add padding
                maxWidth += contentView.paddingRight
                maxHeight += contentView.paddingBottom
                
                setContentSize(maxWidth, maxHeight)
                
            } finally {
                isUpdatingContentSize = false
            }
        }
    }
    
    /**
     * Set explicit content size from Dart side - VirtualizedList approach
     * Matches iOS setExplicitContentSize() behavior
     */
    fun setExplicitContentSize(width: Int, height: Int) {
        explicitContentSize = Pair(width, height)
        setContentSize(width, height)
    }
    
    /**
     * Set content size and notify Dart side
     * Updates the content container's size so ScrollView knows how much to scroll
     */
    private fun setContentSize(width: Int, height: Int) {
        val contentView = getChildAt(0) as? FrameLayout
        if (contentView != null) {
            val layoutParams = contentView.layoutParams
            if (layoutParams != null) {
                if (layoutParams.width != width || layoutParams.height != height) {
                    layoutParams.width = width
                    layoutParams.height = height
                    contentView.layoutParams = layoutParams
                    
                    // Request layout to apply changes
                    contentView.requestLayout()
                }
            }
        }
        
        // Notify Dart side of content size update
        notifyContentSizeUpdate(width, height)
    }
    
    /**
     * Notify Dart side of content size updates through propagateEvent
     * Matches iOS notifyContentSizeUpdate() behavior
     */
    private fun notifyContentSizeUpdate(width: Int, height: Int) {
        // Only propagate event if there are registered listeners
        val eventTypes = getTag(DCFTags.EVENT_TYPES_KEY) as? Set<String>
        if (eventTypes != null && (eventTypes.contains("onContentSizeChange") || eventTypes.contains("contentSizeChange"))) {
            propagateEvent(this, "onContentSizeChange", mapOf(
                "contentSize" to mapOf(
                    "width" to width.toDouble(),
                    "height" to height.toDouble()
                )
            ))
        }
    }
    
    override fun onConfigurationChanged(newConfig: android.content.res.Configuration) {
        super.onConfigurationChanged(newConfig)
        // Clear explicit content size on configuration change (orientation, etc.)
        explicitContentSize = null
    }
}
